velox/functions/prestosql/StringFunctions.h

```diff
@@ -494,4 +496,111 @@ struct LevenshteinDistanceFunction {
   }
 };
 
+/// Wrap the sbstemmer library and use its sb_stemmer_stem
+/// to get word stem
+class Stemmer {
+ private:
+  sb_stemmer* sbstemmer;
+  Stemmer(sb_stemmer* stemmer) : sbstemmer(stemmer) {}
+
+ public:
+  ~Stemmer() {
+    sb_stemmer_delete(sbstemmer);
+  }
+
+  /// Get a Stemmer from the the map stored in thread local storage
+  /// or create a new one if it doesn't exist. Return NULL if the
+  /// specified lang is not supported.
+  static Stemmer* getStemmer(const char* lang) {
+    thread_local std::map<std::string, Stemmer*> stemmers;
```

Comment:
I don't clean up this thread-local map which contains the Stemmer instances for each language. I assume a working thread is shutting down with the Presto process. Not sure if it's worth adding the cleanup code while tearing down a thread.

Reply:
I'm not sure if it is a good idea to use thread local variable here. CC: @bikramSingh91

Reply:
https://github.com/zvelo/libstemmer

```
Creating a stemmer is a relatively expensive operation - the expected
usage pattern is that a new stemmer is created when needed, used
to stem many words, and deleted after some time.

Stemmers are re-entrant, but not threadsafe.  In other words, if
you wish to access the same stemmer object from multiple threads,
you must ensure that all access is protected by a mutex or similar
device.
```

Based on this information, it seems we could add a member variable to 'WordStemFunction' to store Stemmer instances. 

Let's use std::unique_ptr to make cleanup automatic.

Similar to regex functions, we may consider only supporting constant  values for 'lang' and create a single Stemmer instance in 'initialize'.

Alternatively, we may allow up to <small number> of different languages per function instance. See https://facebookincubator.github.io/velox/functions/presto/regexp.html

Reply:
> Based on this information, it seems we could add a member variable to 'WordStemFunction' to store Stemmer instances.

that's the reason I use thread-local to store the lang<-->stemmer map

> Let's use std::unique_ptr to make cleanup automatic.

let me change the map to `<string, unique_ptr<Stemmer>>`

Reply:
To add more context against using thread_local variables liberally is that we observed that storing thread local variables inside objects that can be moved between threads can cause an unexpected increase in memory usage as whenever they are accessed in another thread, they would be created and would exist for the lifetime of the thread. eg of an issue we recently fixed: https://github.com/facebookincubator/velox/pull/7646

In this case as well, the driver executing expression eval (and this code) can bounce between multiple available threads (we use folly::executor to execute drivers), and each time it would end up creating new instances. Unlike #7646, I do not expect this hold on to quite as much memory but nevertheless it would be a good idea to avoid this kind of usage pattern unless there is no other alternative. Using a unique_ptr as @mbasmanova suggested would help us to have tighter control over its lifetime by tying it to the lifetime of the Expression and therefore more predictable memory usage.